pyfarm/agent/http/core/resource.py

```diff
@@ -87,40 +79,54 @@ def template(self):
             raise NotImplementedError("You must set `TEMPLATE` first")
         return template.load(self.TEMPLATE)
 
-    @property
     def methods(self):
-        """A set containing all the methods this resource implements."""
-        methods = set()
-        for method in ("get", "post", "put", "delete", "head"):
-            attribute_count = 0
-            for attribute_name in (method, "render_%s" % method.upper()):
-                attribute = getattr(self, attribute_name, None)
-                if attribute is not None:
-                    attribute_count += 1
-                    methods.add(method)
-
-            if attribute_count == 2:
-                raise ValueError(
-                    "%s has both `%s` and `%s` methods" % (
-                        self.__class__.__name__,
-                        method, "render_%s" % method.upper()))
-
-        return methods
-
-    def content_types(self, request, default=None):
-        """Returns the content type(s) present in the request"""
-        headers = request.requestHeaders.getRawHeaders("content-type")
-        if isinstance(default, STRING_TYPES):
-            default = [default]
+        """
+        Returns a tuple of methods which an instance of this class implements
+        """
+        methods = []
+        for method_name in ("get", "put", "post", "delete", "head"):
+            method = getattr(self, method_name, None)
+            if method is not None and callable(method):
+                methods.append(method_name)
+
+        return tuple(methods)
+
+    def response_types(self, request, default=None):
+        """
+        Returns an instance of :class:`frozenset` which contains the type
+        or types of responses we can send.  We look at the ``Accept`` and
+        ``Content-Type`` headers to determine this information.  If values
+        are present in the ``Accept`` header these will be returned instead
+        of values in ``Content-Type``.  If neither of these headers are present
+        we'll return what was provided to ``default`` instead.
```

Comment:
The ```Content-Type``` header is NOT supposed to be used in this way.

Quote from the RFC:

     If no Accept header field is present, then it is assumed that the client accepts all media types.